### ISSUE TITLE
Could we get a contribution/style guideline and Makefile?

### ISSUE THREAD

**Issue by Shatnerz** [rust-bitcoin/rust-bitcoin#685]
This might lower the barrier to entry for outside patches.

I'm new to the project, but basically what I have in mind is a simple Makefile that spells out a few basic commands like

- format
- clippy
- different test groups

Update the Pull Request section in the README to inform people about formatting for new code. Maybe add a CI job is possible

**Comment by apoelstra** [comment]
The formatting for new code, like any open-source project, is "copy the existing formatting style".

If you want to run clippy from time to time and clean stuff up, that would be welcome, but Clippy historically has flagged a ton of completely non-problematic code (one that sticks out to me is that it didn't like us implementing a bignum subtraction with non-subtraction operators..) and it does not make sense to try to run it in CI.

I'm not sure what you mean by "different test groups" or by "Makefile". It definitely would be an improvement if people understood how to run a broad set of tests at home. Maybe we should have a script that, say, runs the unit tests and examples with various random feature combinations.

**Comment by Shatnerz** [comment]
> The formatting for new code, like any open-source project, is "copy the existing formatting style".

The issue is that often no style is explicitly defined and it has to be implicitly learnt from the codebase. As contributions come in, the author may simply be patching one section of the code and they won't be familiar with an overarching style. Overtime style can diverge within a single project leading to confusion and increasing the barrier to entry for a contribution.

> I'm not sure what you mean by "different test groups" or by "Makefile". It definitely would be an improvement if people understood how to run a broad set of tests at home. Maybe we should have a script that, say, runs the unit tests and examples with various random feature combinations.

Basically what you suggested. Something simple so a dev just needs to run

- `make format`
- `make clippy`
- `make unit-test`
- `make {feature}-test`
- `make ci-test`

or something like that. I'm new to the codebase and not too familiar but a bare minimum with a style check and unit tests would be nice. Not all those examples may be relevant.

**Comment by Kixunil** [comment]
@apoelstra I believe that the benefit of running clippy outweighs the cost of adding a few `#[allow(clippy::foo)]` statements.
If all confirmed-OK warnings are `#[allow(...)]`ed then it also makes sense to run it in CI. Unless there's something with a huge amount of FPs but in that case I'm sure clippy maintainers would appreciate a report.

@Shatnerz I don't think Makefile provides significant value esp. given Makefiles are super-annoying to write. There isn't much difference between e.g. `make clippy` and `cargo clippy`. Script to run multiple tests is probably the only thing that seems useful to me but that doesn't need a Makefile.

Spelling out the commands in CONTRIBUTING.md may be a good idea if we agree on them.

**Comment by apoelstra** [comment]
I kinda like the idea of using `make` to do a bunch of things at once, but agreed that a shellscript would probably be clearer and we probably don't need any of the fancy dependency-solving features of `make`.

I understand the concerns about code formatting but AFAIK there are no rust formatting tools out there that are acceptable to all maintainers of this crate.

**Comment by Shatnerz** [comment]
Yeah, I wouldn't get hung up on `make`. I only mentioned it because I've seen a few private projects use it with success. I checked a bunch of popular crates and it doesn't seem that popular. I'm personally not a fan of `make`. The main takeaway is a single location to simplify common actions. Someone coming in with 0 context can see a `Makefile` or `run.sh` (or whatever is an appropriate name) and can see how to lint, format, unit test, integration test, test specific feature, etc...

> Spelling out the commands in CONTRIBUTING.md may be a good idea if we agree on them.

Yeah, something like this would be nice. It pairs nicely with formatting the code if a style or format can be agreed on.

> I understand the concerns about code formatting but AFAIK there are no rust formatting tools out there that are acceptable to all maintainers of this crate.

Don't hold me to this, but I might be able to try some things out and come up with the least restrictive formatting/linting. If that can at least be defined, the style shouldn't diverge much. If you'd want to make things more restrictive you could update the rules and push the changes.

**Comment by Shatnerz** [comment]
@apoelstra ah yes, that would be annoying. I'll update this issue and link a PR if I manage anything

**Comment by dpc** [comment]
Introducing `rustfmt` has an open issue already #172

From my perspective a lot of projects I worked with have a trivial `Makefile` where the commands are just `@./scripts/dosomething.sh` or `@cargo build` etc. The reason for it is that it's a language-independent convention that introduces and documents all the common things a newcomer might want to be aware of. A default `make` should possibly print available commands and some links to "getting started" documentation. IMO, it's a much better convention than hoping people will read everything in the documentation. The documentation can just say "run `make` for help with getting started with the project".

In my future projects I'm planning to migrate to using https://github.com/casey/just which makes a lot of this even easier, with stuff like `just -l`, at the cost of being a newer and less popular tool.

**Comment by Kixunil** [comment]
I'm in Rust for almost as long as it's stable and I don't remember seeing a single crate with a Makefile. (Maybe rustc itself had one?)

For people who know Rust but don't know the `bitcoin` crate, the lack of Makefile is completely normal and they already know to run `cargo clippy`, `cargo test` and probably would attempt `cargo fmt` and quickly find that it changes too much stuff and do `git reset --hard HEAD`.

For people who don't know Rust it's best to learn standard Rust practices first. Don't want to alienate anyone and it's great that you @Shatnerz are trying to learn, I just don't think putting a Rust tutorials to every crate scales. If there are important things missing in some official/popular docs that's probably a problem for wider community to solve. If there is some resource that is better than The Book I'm fine with linking it.

Also, we should finally finish #587 and maybe add some TL;DR there.

**Comment by Shatnerz** [comment]
> I just don't think putting a Rust tutorials to every crate scales

It's not about a tutorial. It's about standardizing style and format and lowering the barrier-to-entry for contributions. For example on current `master` (a961ab45)
- `cargo clippy` has 113 warnings
- `cargo fmt` generates a massive diff
- `cargo test` excludes major feature sets which needs to be searched for in order to run (I suppose this is normal though)

I suppose it's clear that these aren't used. Now what? Either a contribution forks the existing style, someone ends up repeating stylistic guidelines anytime something new emerges, or contributions will be closed for not match the implicit style.

It's nice to see #587 in the works and we can likely close this as a duplicate, but the current docs add little information about actually contributing

> Every PR needs at least two reviews to get merged. During the review phase maintainers and contributors are likely to leave comments and request changes. Please try to address them, otherwise your PR might get closed without merging after a longer time of inactivity. If your PR isn't ready for review yet please mark it by prefixing the title with WIP: .

Leading much to be inferred. I'll also add the the CI pipeline isn't obvious and quite easy to miss as well.

In the end @apoelstra's comments make it clear that automated linting and formatting are unlikely.

If I don't try any changes in the next 10-14 days, feel free to close this as a stale issue.

**Comment by Kixunil** [comment]
Yeah, there's a lot to be improved. I fixed most clippy lints in #686 to push this forward.

**Comment by sanket1729** [comment]
I think there was a rough consensus for `rustfmt` #171  and there was also a rough agreement on introducing it file by file. I don't think there is anyone actively working on it. 

I like Clippy, but based on my experience dealing with Clippy for every PR in CI might be annoying.  I think we should address `rustfmt` first. Meanwhile, PRs like #686 are welcome.

**Comment by Kixunil** [comment]
Interesting, I would've expected clippy to be more bearable because there are generally fewer warnings.

**Comment by thomaseizinger** [comment]
Relevant prior art: https://github.com/matklad/cargo-xtask

**Comment by apoelstra** [comment]
Sorry to necro this but I think putting a Makefile in which just echos "To build run `cargo build`. To run the unit tests run `cargo test`. To run specific test run `cargo test <test name>`. Would be low-effort and super helpful.

I agree we can't provide a full rust tutorial but for people downloading the library who have never used Rust before it will at least give them some commands to run (or figure out how to install) so they can get off the ground.

As one data point, I have almost never successfully used Go or Java because of this sort of startup friction. I just switch to an alternative written in a language whose build system I am familiar with.

**Comment by Kixunil** [comment]
Wouldn't it make more sense to add these instructions to `README.md` or `BUILD.md`? `Makefile` has certain semantics and could confuse some automated tools. I'd really prefer avoiding it.

I find it strange that complete newbies would start contributing to libraries as opposed to binaries, which are generally more interesting. Maybe there's some nice binary using rust-bitcoin we could suggest newbies to look at?

@Shatnerz could you explain your thought process of deciding to look at rust-bitcoin? We should be able to improve the situation better if we understand it.

**Comment by Shatnerz** [comment]
@apoelstra 

> but I think putting a Makefile in which just echos "To build run cargo build. To run the unit tests run cargo test. To run specific test run cargo test <test name>. Would be low-effort and super helpful.

So I'm pretty sure I opened this after I found running the CI tests locally a pain initially. I think that was half of the motivation here. It would also be nice have something that maybe runs through the tests with either specific/all feature flags enabled, or a mix of specific important features if features can potentially collide and cause unwanted behaviour.

The other half of the motivation was minor comments like:
- [style comment](https://github.com/rust-bitcoin/rust-bitcoin/pull/684#discussion_r738815871)
- [suggestion to use clippy](https://github.com/rust-bitcoin/rust-bitcoin/pull/684#discussion_r739266704)

Great, but are people using it or do they know to use it (basically [this message](https://github.com/rust-bitcoin/rust-bitcoin/issues/685#issuecomment-958425245) If so, why is `cargo clippy` returning so many warnings? Why does `cargo fmt` generate a massive diff? I won't be able to format my PR if it creates a diff across the entire project.




> Wouldn't it make more sense to add these instructions to README.md or BUILD.md? Makefile has certain semantics and could confuse some automated tools. I'd really prefer avoiding it.

Yes, this would be preferable. This requires maintaining the README which can often fall behind.

> @Shatnerz could you explain your thought process of deciding to look at rust-bitcoin? We should be able to improve the situation better if we understand it.

This suggestion was spawned from the fact that recreating the CI tests locally was a pain in the ass. I created https://github.com/rust-bitcoin/rust-bitcoin/pull/684 and tests passed locally running `cargo test`. Great, I pushed my code, wait for a CI approval, and then run into failures and have to work backwards from there and make some inferences based on why

**Comment by Shatnerz** [comment]
@Kixunil

> Wouldn't it make more sense to add these instructions to README.md or BUILD.md?

Yes, but it is common for READMEs to fall behind. If devs and the CI are running the same sets of scripts, it's easier to infer what is happening and one doesn't need to worry about stale READMEs.

> Makefile has certain semantics and could confuse some automated tools. I'd really prefer avoiding it.

See [this comment](https://github.com/rust-bitcoin/rust-bitcoin/issues/685#issuecomment-958075322). Basically don't get hung up on the idea of a Makefile. A shellscript or similar could accomplish the same. Anything that is easy to logically find at the root of the project.

> I find it strange that complete newbies would start contributing to libraries as opposed to binaries, which are generally more interesting. Maybe there's some nice binary using rust-bitcoin we could suggest newbies to look at?

I don't assume you had negative intentions here, but this comment may come off as a tad aggressive some. I'd use softer language in the future to avoid scaring people off or derailing conversations. Anyways, I have worked in the crypto space for several years now. I thought "while not mix in what I know to help me learn Rust". I start playing around with some ideas I had. I was using `rust-bitcoin` when I discovered that [`is_signed_by_address` isn't implemented for all but 1 address type yet doesn't return a `NotYetImplemented` error](https://github.com/rust-bitcoin/rust-bitcoin/blob/fe43e3c9d764696f720c73e4444cfe9377d4114a/src/util/misc.rs#L166-L175). It's within my domain, it's clearly wrong/problematic, and it's in `util/misc.rs` so it likely isn't a core piece of the lib. It was a great opportunity for a patch to support the library and in return to have my code critiqued by those more familiar than the language than I am.

**Comment by Kixunil** [comment]
@Shatnerz thanks for valuable input!

> Basically don't get hung up on the idea of a Makefile. A shellscript or similar could accomplish the same. 

Yeah, this exactly. I agree with a shell script, just *not Makefile specifically*.

> I don't assume you had negative intentions here, but this comment may come off as a tad aggressive some.

Yeah, maybe I should word it as "I'm surprised that people would contribute to libraries". My intention was to say that AFAIK newbies usually try simpler things first, so hopefully they would know some basic stuff already.

>  I was using `rust-bitcoin`

I assume in some kind of binary? That's what I was talking about, basically.
The idea is that if you weren't informed about e.g. clippy existing it may be a fault of Rust documentation or whatever sources you picked to learn, so it'd be more useful/efficient to fix that than having a Rust guide in every Rust project. I wouldn't mind linking recommended learning resource though.

**Comment by Kixunil** [comment]
This discussion inspired me to write this: https://gist.github.com/Kixunil/ba491d01c17eecfe92de992b2fe04a73
May be worth linking?

**Comment by tcharding** [comment]
Thanks for your input @Shatnerz.

re: Clippy

We've only got 31 Clippy warnings left, many of these will go after we bump MSRV and switch to edition 2018. More than one maintainer is acutely aware of the Clippy situation and actively working on improving it and getting us to a stage that we can run Clippy on CI.

re: Contributor documentation

We recently merged a `CONTRIBUTING.md` document.

Can we close this issue @Shatnerz?
